/test: use a different image for hostname override e2e tests

neolit123 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

As discovered in ticket 134737 `hostname` is buggy on busybox
due to the musl backend that it uses. something in the /etc/hosts
that k8s generates trips its parser and it doesn't work properly
in the ipv6 tests.

To workaround that use an image that has a glibc backend,
so that the hostname command works.

#### Which issue(s) this PR is related to:
<!--
Please link relevant issues to help with tracking.

To automatically close the linked issue(s) when this PR is merged,
add the word "Fixes" before the issue number or link.
Do not use "Fixes" if the PR is of kind `failing-test` or `flake`.

Reference KEPs when applicable in addition to specific issues.

Examples:
Fixes #<issue number>
<issue link> (issue in a different repository)
KEP: https://github.com/kubernetes/enhancements/issues/<kep-issue-number>

If there is no associated issue, then write "N/A".
-->

fixes https://github.com/kubernetes/kubernetes/issues/134737

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

neolit123 commented:
/test pull-kubernetes-e2e-kind-ipv6

neolit123 commented:
/retest

>E2eNode Suite: [It] [sig-node] MirrorPod when create a mirror pod without changes should successfully recreate when file is removed and recreated [NodeConformance] expand_more	2m4s
e2e.go: Node Tests expand_more

neolit123 commented:
/test pull-kubernetes-e2e-kind-ipv6

neolit123 commented on test/e2e/common/node/pod_hostnameoverride.go:
```diff
@@ -44,7 +44,7 @@ func newTestPod(namespace string) *v1.Pod {
 					Name:    "test-pod-hostname-override",
 					Image:   imageutils.GetE2EImage(imageutils.Agnhost),
-					Command: []string{"sh", "-c", "echo $(hostname)';'$(hostname -f)';'"},
+					Command: []string{"sh", "-c", "while IFS= read -r line; do echo \"$line\"; done < /etc/hosts"},
```
do we need to branch for OS here and do this on windows?
```
Command: ["cmd", "/c", "type c:\\windows\\system32\\drivers\\etc\\hosts"]
```

neolit123 replied:
the KEP states:

> For Windows containers, we only set the container's hostname and do not create an /etc/hosts file for it (as we have previously made it clear that we do not create an /etc/hosts file for Windows containers).

so i would assume the feature HostnameOverride is not supported on Windows as it doesn't amend the Windows hosts file (which isn't /etc/hosts)?

https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/4762-allow-arbitrary-fqdn-as-pod-hostname/README.md

EDIT: or i may be misunderstanding here.
- the windows container receives an override hostname
- there is no /etc/hosts
- calling `hostname -f` from aghhosts works (somehow) so the e2e test worked on windows.

if that is the case we should be likely calling a Windows native way to get the hostname instead, like `hostname.exe`, but there is no `hostname.exe -f` support, AFAIK.

aojea replied:
let's switch just the image https://github.com/kubernetes/kubernetes/pull/134757/files#r2450537200

neolit123 replied:
that is understand as well. historically the hosts file under system32 has been a  unix-like artifact, and the format is the same.

neolit123 replied:
i guess this highlights the point @SergeyKanzhelev brought up about the linux vs windows feature discrepancy. i personally think that a sig-windows +1 should be a blocker for all sig-node keps reviews, which i understand is a bandwidth problem too.

alt, the project could establishes a policy that "all new kubelet features are unsupported on windows, unless sig-windows approved and opts-in", that might be easier. and if a feature is desired by windows users, they can contribute changes and also amend keps and e2e...which can ease the burden on sig-windows.

aojea commented on test/e2e/common/node/pod_hostnameoverride.go:
```diff
@@ -44,7 +44,7 @@ func newTestPod(namespace string) *v1.Pod {
 				{
 					Name:    "test-pod-hostname-override",
 					Image:   imageutils.GetE2EImage(imageutils.Agnhost),
```
```suggestion
					Image:   imageutils.GetE2EImage(imageutils.JessieDnsutils),
```

Just that, I do not feel is great to rely on parsing the etc hosts for this test, hostname exercises the userspace calls that lot of languages use and we found this bug in musl thanks to that ... so is not about passing the test is about detecting this incompatibilities

neolit123 replied:
i will switch the image. the questions from above about windows support stand.

HirazawaUi replied:
Alternatively, we could also use `SkipIfNodeOSDistroIs` to skip the test when the node is a Windows node. However, the question remains whether we already actually use this logic now, considering that the same test exists in https://github.com/kubernetes/kubernetes/blob/1404757c5a8194659ddbc5e14a3d8a143d11086a/test/e2e_node/pod_hostnamefqdn_test.go, and so far no one has raised any issues about it not being suitable for Windows.

HirazawaUi replied:
I'm not sure if we currently have any windows-based e2e test jobs running. However, there is an ongoing PR discussing the `LinuxOnly` label for e2e test cases:
ref: https://github.com/kubernetes/kubernetes/pull/134413

No conclusion has been reached yet. If you're concerned about this test case failing on Windows, we could potentially add the `[LinuxOnly]` label to the current test case for now, until we determine the appropriate labeling strategy for such tests in the #134413 discussion.

BenTheElder replied:
I don't know how current it is but I just triggered one, you can find all jobs with `/test` (no parameter) or else the config in test-infra.

I need to catch up on this linuxonly thread, I had not seen that one yet.

neolit123 committed:
/test: use a different image for hostname override e2e tests

As discovered in ticket 134737 `hostname` is buggy on busybox
due to the musl backend that it uses. something in the /etc/hosts
that k8s generates trips its parser and it doesn't work properly
in the ipv6 tests.

To workaround that use an image that has a glibc backend,
so that the hostname command works.

neolit123 commented:
updated with the jessy image.

/test pull-kubernetes-e2e-kind-ipv6

BenTheElder commented:
/test pull-kubernetes-e2e-capz-windows-master
/test pull-kubernetes-e2e-kind-ipv6

marosset commented:
It looks like the windows job passed
https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/134757/pull-kubernetes-e2e-capz-windows-master/1981072751577796608 I'm not sure why the prow job is reporting it failed :(

neolit123 commented:
IIUC, we are not testing this feature on windows (at least in this capz job) due to the ginkgo flags there.

https://github.com/kubernetes/kubernetes/pull/134757#discussion_r2453167822

BenTheElder reviewed (commented):
/lgtm
/approve
